streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java

```diff
@@ -16,349 +16,173 @@
  */
 package org.apache.kafka.streams.processor.internals.assignment;
 
+import org.apache.kafka.common.protocol.ByteBufferAccessor;
 import org.apache.kafka.streams.errors.TaskAssignmentException;
+import org.apache.kafka.streams.internals.generated.SubscriptionInfoData;
 import org.apache.kafka.streams.processor.TaskId;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.util.Collection;
-import java.util.HashSet;
 import java.util.Set;
 import java.util.UUID;
-
-import static org.apache.kafka.streams.processor.internals.assignment.StreamsAssignmentProtocolVersions.LATEST_SUPPORTED_VERSION;
-import static org.apache.kafka.streams.processor.internals.assignment.StreamsAssignmentProtocolVersions.UNKNOWN;
+import java.util.stream.Collectors;
 
 public class SubscriptionInfo {
-    private static final Logger log = LoggerFactory.getLogger(SubscriptionInfo.class);
-
-    private final int usedVersion;
-    private final int latestSupportedVersion;
-    private UUID processId;
-    private Set<TaskId> prevTasks;
-    private Set<TaskId> standbyTasks;
-    private String userEndPoint;
 
-    // used for decoding; don't apply version checks
-    private SubscriptionInfo(final int version,
-                             final int latestSupportedVersion) {
-        this.usedVersion = version;
-        this.latestSupportedVersion = latestSupportedVersion;
-    }
+    private static final Logger log = LoggerFactory.getLogger(SubscriptionInfo.class);
 
-    public SubscriptionInfo(final UUID processId,
-                            final Set<TaskId> prevTasks,
-                            final Set<TaskId> standbyTasks,
-                            final String userEndPoint) {
-        this(LATEST_SUPPORTED_VERSION, processId, prevTasks, standbyTasks, userEndPoint);
+    public static final int LATEST_SUPPORTED_VERSION = SubscriptionInfoData.SCHEMAS.length - 1;
+    static final int UNKNOWN = -1;
+    private final SubscriptionInfoData data;
+
+    private static void validateVersions(final int version, final int latestSupportedVersion) {
+        if (latestSupportedVersion == UNKNOWN && (version < 1 || version > 2)) {
+            throw new IllegalArgumentException(
+                "Only versions 1 and 2 are expected to use an UNKNOWN (-1) latest supported version. " +
+                    "Got " + version + "."
+            );
+        } else if (latestSupportedVersion != UNKNOWN && (version < 1 || version > latestSupportedVersion)) {
```

Comment:
don't we use a number higher than the `latestSupportedVersion` for version probing?

Reply:
Yeah, it's confusing...

The message itself contains both "my used version" and "my latest supported version" fields. This is validating that the sender never reports that it's using a version higher than its own latest supported version, which would be nonsense.

There's also _this instance's_ (i.e., the receiver of the message) latest supported version. The version probing case is when the sender is using a version higher than the _receiver's_ latest supported version.